### DISCUSSION TITLE
AWS Batch: What happens after `maxTransferAttempts`?

### DISCUSSION THREAD

**Discussion by arareko** [nextflow-io/nextflow#3722] [Q&A] [answered]
Hi,

We're using `21.08.0-edge` and have a configuration with the following parameters for AWS Batch:

```
aws {
    batch {
        // Retry S3 downloads if they fail, waiting 20 seconds to retry
        maxTransferAttempts = 5
        delayBetweenAttempts = 20

        // Lower I/O load to try and prevent failures with multiple jobs
        // on a single instance
        maxParallelTransfers = 5
    }
}
```

I'd like to know what happens after the specified number of attempts is exhausted? For our case, the pipelines continue to run even when downloads were incomplete, obviously producing inconsistent results. Does Nextflow have a configuration/option for triggering a fatal error and stopping the pipeline when these things happen?

**Comment by mribeirodantas** [comment]
I'm not fully aware of the AWS Batch plugin for Nextflow. Having said that, you could create a process to check if all files were appropriately downloaded and `throw` an exception if they are not. The `throw` will terminate the pipeline.

**Comment by delagoya** [comment]
NF uses the  AWS CLI to download files via a Bash script that is created on the fly, and these options are passed to it. 

I would have to dig into the code to see if the exit status is captured from that Bash script.

**Reply by arareko** [reply]
@delagoya Are these the same options that were fixed in [this commit](https://github.com/nextflow-io/nextflow/commit/f1a58600d063bb217afa01ca949bc9299ee33828), which was added after the version of NF that we currently use?

**Reply by pditommaso** [reply]
Uploads and downloads and executed in parallel by the snippet below. If the transfer fail, the error code is ignored however the corresponding that will fail for the lack of the expected file.

https://github.com/nextflow-io/nextflow/blob/b93634e900d4e58d529e297a486d67ae687f5b2c/modules/nextflow/src/main/groovy/nextflow/executor/BashFunLib.groovy#L84-L113

Said that, S3 CLI approach has serious scalability issues. Recently we have introduced the [Fusion file system](https://seqera.io/blog/breakthrough-performance-and-cost-efficiency-with-the-new-fusion-file-system/) to tackle this problem

**Reply by arareko** [reply, marked as answer]
@pditommaso @delagoya Thanks for the replies and the link to the Fusion file system blog post, it looks great and really worth investigating updating all of our pipelines to make use of it.

With regards to the original issue we were experiencing, we were able to fix after updating to a slightly newer version of Nextflow than the one we had (`21.08.0-edge` -> `21.10.6-all`), which contains the commit [I mentioned before](https://github.com/nextflow-io/nextflow/discussions/3722#discussioncomment-5270317) and not a lot more (for now we only wanted to introduce very little change to not break other things).
